# **Squence in Python**


In Python, a **Sequence** is a generic term for a data structure that holds multiple ordered items, where each item can be accessed directly using an integer position number known as an **Index**. 

Think of a sequence like a train: it has separate compartments chained together in a strict, predictable row, and every compartment has a specific coach number.

# **length  squence in python**
len()

In [ ]:

a = [768,7.8,"ali"]
print(len(a))
b = "Squences"
print(len(b))
c = (9,44,20)
print(type(c))
print(len(c))


3
8
<class 'tuple'>
3


# **membership squences in python**
use of in    or not in

In [ ]:

# use of in    or not in
bazar = ["fruit","mall","shops"]
print("fruit" in bazar)
print("saloon" not in bazar)
print("ali" in bazar)
tuck_shop = ("bisket","juice","ice cream","water botal")
print("bisket" in tuck_shop)
print("Juice" in tuck_shop)  # output : false    because it case sensitive
print("juice" in tuck_shop)
print("car"not  in tuck_shop)

True
True
False
True
False
True
True


# **Concatination  squences in python**


Use of    **+**

In [17]:
first_name = "Muhammad "
second_name = "Farhan"
print(first_name + second_name)  # string + string 
# list + tuple = X
# same + same = correct  execution
product= ["potatoes","ginger","mangoes"]
price = [120,900,500]
print(product + price)

books = ["english","mathamatics"]
new_books = ["islamiyat","social science" ,"Artificial intellegence"]
print(books + new_books)

Muhammad Farhan
['potatoes', 'ginger', 'mangoes', 120, 900, 500]
['english', 'mathamatics', 'islamiyat', 'social science', 'Artificial intellegence']


# **Repitition of squenceces in python**
use of *

In [25]:
a = 6
print(a *5)
print("________" *5)
b = "ali"
print(b*6)
c = ("ohh","hooo")
print(c*7)
d = (8,98)
print((d,)*5)


30
________________________________________
alialialialialiali
('ohh', 'hooo', 'ohh', 'hooo', 'ohh', 'hooo', 'ohh', 'hooo', 'ohh', 'hooo', 'ohh', 'hooo', 'ohh', 'hooo')
((8, 98), (8, 98), (8, 98), (8, 98), (8, 98))


# **Range squences in python**

In [31]:
a =(30)
print(range(a,50,5))
print(list(range(a,50,5)))

gen =(20)
print(range(gen,121,10))
print(list(range(gen,121,10)))

range(30, 50, 5)
[30, 35, 40, 45]
range(20, 121, 10)
[20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120]


# **Indexing squence in Python**

**In String**

In [36]:
a = "pakistan"
print(a[1])  # positve start from 0
print(a[-1])  # negative start from -1

b = "what is the defination of range"
print(b[9])
print(b[-6])   # space also consider as a index number and print as empty 

a
n
h
 


**in list**

In [37]:
a = [7,9,0,8,9,4,6,0]
print(a[-1])
print(a[3])

0
8


**IN Tuple**

In [40]:
a = (8,6,4,6,7)
print(a[-1])
print(a[4])

7
7


# **Nested Structure in python indexing**

In [47]:
a = ["ali","ahmad",3],["usman","umar",90]
print(a[1][0])  # here are two sub list its like    [[],[]]  
# [ ]this is main list and 
#[][] these are sub list

b = [["ali","arman","python"],["java","c++","replit"]]
print(b[0][1])
print(b[1][-1])

usman
arman
replit


# **Slicing in python**

In [64]:
a = [1,2,3,4,5,6,7]
print(a[1:-2])

b = 10
print(range(b,550,10))
print(list(range(b,550,10)))
list_range = list(range(b,550,10))
print(list_range[1:-2:5])

[2, 3, 4, 5]
range(10, 550, 10)
[10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130, 140, 150, 160, 170, 180, 190, 200, 210, 220, 230, 240, 250, 260, 270, 280, 290, 300, 310, 320, 330, 340, 350, 360, 370, 380, 390, 400, 410, 420, 430, 440, 450, 460, 470, 480, 490, 500, 510, 520, 530, 540]
[20, 70, 120, 170, 220, 270, 320, 370, 420, 470, 520]


# 🔬 Advanced Slicing Logic Breakdown

Given the syntax: `list_range[1:-2:5]`

### The 3-Step Execution Path:
1. **Start at Index 1:** Captures the value `20` as the baseline starting point.
2. **Set the Stop Boundary (-2):** Targets index `-2` (value `530`). Slicing stops safely right before this boundary, making `520` the highest possible element.
3. **Apply Step Increment (5):** Jumps forward by 5 positions dynamically.

### Step-by-Step Index Map:
* Index `1`  ➔ **`20`**
* Index `6`  ➔ **`70`**
* Index `11` ➔ **`120`**
* Index `16` ➔ **`170`**
* Index `21` ➔ **`220`**
* Index `26` ➔ **`270`**
* Index `31` ➔ **`320`**
* Index `36` ➔ **`370`**
* Index `41` ➔ **`420`**
* Index `46` ➔ **`470`**
* Index `51` ➔ **`520`** *(Stops here; next jump breaks the boundary limit)*


# **Modifying the element using indexing**

mutable means changing something from the data not fully data only some features

# **Modifying List**

In [ ]:
a = ["ali","arman","usman"]
print(a) # before 
a[1] = "hamza"
print(a) # after

['ali', 'arman', 'usman']
['ali', 'hamza', 'usman']
